### PR TITLE
Stream stop event

### PR DESCRIPTION
**Pull request by silverbucket**
When we receive a stream stop we should propagate an event for client code.
https://datatracker.ietf.org/doc/html/draft-ietf-quic-transport-29#page-127

### CODE REVIEW

**Comment by silverbucket** [comment]
Coverage and tests passing, not sure what to do about the package-wheel build error.

**Comment by silverbucket** [comment]
`cibuildwheel` doesn't support versions under python3.6, however it's failing on python 3.10, could be a version matching bug.

**Comment by silverbucket** [comment]
@jlaine - It looks like the reason for the failure is that the check for python `< 3.6` is incorrectly including `3.10` and therefore thinks it's a version lower than `3.6`. 

Have you had a look at this PR otherwise? The code and coverage itself should be ready for review.

File: tests/test_connection.py [hunk 1]
```diff
@@ -1734,17 +1734,28 @@ def test_handle_retire_connection_id_frame_invalid_sequence_number(self):
...
-            # client receives STOP_SENDING
+            # client receives STOP_STREAM
             client._handle_stop_sending_frame(
                 client_receive_context(client),
                 QuicFrameType.STOP_SENDING,
-                Buffer(data=b"\x00\x11"),
+                Buffer(
+                    data=encode_uint_var(stream_id)
+                    + encode_uint_var(QuicErrorCode.INTERNAL_ERROR)
+                    + encode_uint_var(0)
```
[the remark this answers is not in the document]
Comment (silverbucket) [reply tests/test_connection.py:1750]: @jlaine This is defined in the spec as the application protocol error code. 
https://datatracker.ietf.org/doc/html/draft-ietf-quic-transport-29#section-19.5

**Comment by silverbucket** [comment]
Hi @jlaine sorry for the delay, this has been on my todo list for quite a while. I've tried to address your feedback, let me know if there's anything else I may not have addressed.

> I do think we need to expose this kind of event through the API. However I'm unsure about a couple of things:
> 
> * naming: neither the name StreamStop nor the doc strings correctly convey the fact that it's only the "sending" part (from our perspective) of the stream which has been stopped

I agree, the naming is not ideal, but nothing came to mind that was more obviously appropriate and also, I was trying as best I could to stick to existing conventions within the code. I've updated the docstring to try to clarify, but open to suggestions as well.

> * if we send a RESET_STREAM frame on a stream, do we want to be notified when it is ACK'd?

This I'm not sure of, is it something that you feel should be taken care of in this PR? If so I would defer to your lead.

**Comment by silverbucket** [comment]
@jlaine any thoughts on this?

**Comment by jlaine** [comment]
My thoughts remain the same: that the API/naming are unclear. Unfortunately I don't have the mental bandwidth for a better proposal right now,..

**Comment by silverbucket** [comment]
@jlaine What about making an issue to address these questions further once you have the capacity, and meanwhile accepting this PR? We've been using it as part of our test infrastructure with the PR+latest master for over 7 months without any issue.

**Comment by jlaine** [comment]
> @jlaine What about making an issue to address these questions further once you have the capacity, and meanwhile accepting this PR? We've been using it as part of our test infrastructure with the PR+latest master for over 7 months without any issue.

Uhm no thanks. Code that gets merged adds to my maintenance burden, doubly so if it means backwards-incompatible changes down the line!

**Comment by jlaine** [comment]
To be clear: I don't mean "I don't want this feature merged", I do think it definitely adds value. What I mean is "I don't want it merged half-baked". Please help me do the hard work of thinking it all the way through. Sorry about the tone of my previous message.

**Comment by vranem1** [comment]
What about name StopSendingReceived ?

**Comment by jlaine** [comment]
Superseeded by #393

**Pull request closed without merging**